modules/windows-hello.nix

```diff
@@ -0,0 +1,50 @@
+{ lib, pkgs, config, ... }:
+
+with builtins; with lib;
+{
+
+  options.wsl.windowsHello = {
+    enable = mkEnableOption "Authentication using Windows Hello";
+  };
+
+  config =
+    let
+      cfg = config.wsl.windowsHello;
+    in
+    mkIf (config.wsl.enable && cfg.enable) {
+
+      security.sudo.wheelNeedsPassword = true;
+      security.sudo.extraConfig = ''
+        Defaults rootpw
+      '';
+
+      # Hijack the pam_usb module, because NixOS does not allow for adding custom PAM modules at the moment
+      security.pam.usb.enable = true;
+      nixpkgs.overlays = [
+        (self: super: {
+          pam_usb =
+            let
+              authenticator = pkgs.stdenv.mkDerivation {
+                name = "WindowsHelloAuthenticator.exe";
+                src = pkgs.fetchurl {
+                  url = "https://github.com/nzbr/PAM-WindowsHello/releases/download/v1/WindowsHelloAuthenticator.exe";
+                  sha256 = "4856a1fefa5c869b78890f9313a560d310e9c11f2a2a212c2868cf292792ff7f";
+                };
+                dontUnpack = true;
+                buildCommand = ''
+                  install -m 0755 $src $out
+                '';
+              };
+              wrapper = pkgs.writeShellScript "wrapper" ''
+                export PATH=${pkgs.coreutils}/bin # The PAM environment does not include the default PATH
+                export WSL_INTEROP="/run/WSL/$(ls -tr /run/WSL | tail -n1)" # Find the correct WSL_INTEROP socket to be able to start the EXE
+                exec ${authenticator} [$PAM_SERVICE] $PAM_RUSER wants to authenticate as $PAM_USER
+              '';
```

Comment:
I think we should make a proper package out of that. Could we build it from source, too?

Reply:
I don't think it'd be possible to build from source. The program has build-time dependencies against `Windows.winmd` and `System.Runtime.WindowsRuntime.dll`
If there's any way to get those files into the nix store, it might be possible to build with Mono though

Reply:
Both are on nuget https://www.nuget.org/packages/Microsoft.Windows.WinMD/ https://www.nuget.org/packages/System.Runtime.WindowsRuntime/

Reply:
I didn't expect that, thanks! I'll look into it

Reply:
I tried it now (the correct package is https://www.nuget.org/packages/Microsoft.Windows.SDK.Contracts btw. The one above seems to be for parsing WinMD files), however I could not get msbuild on Linux to compile the executable. It always threw an error about a missing assembly reference. I also tried porting the helper to .NET 6, but that wouldn't build on Linux either because `NETSDK1100: Windows is required to build Windows desktop applications`. I had initially meant to build it in .NET 6, but couldn't figure out how to import WinRT then. In retrospect that was fortunate, because now I learned that .NET 6 would require having an additional .dll in the same directory as the .exe, whereas .NET Framework only requires the .exe to be present

TL;DR Unfortunately it looks like there is no way to build the helper on linux